pkg/transformer/util.go

```diff
@@ -70,3 +72,148 @@ func unmarshalRequest(data []byte, v interface{}) error {
+// 	- Vin[i].N (vout number) -> get Transaction(txID).Vout[N].Address
+// 	- returning address already has 0x prefix
+func getNonContractTxSenderAddress(p *qtum.Qtum, vins []*qtum.DecodedRawTransactionInV) (string, error) {
+	for _, vin := range vins {
+		prevQtumTx, err := p.GetRawTransaction(vin.TxID, false)
+		if err != nil {
+			return "", errors.WithMessage(err, "couldn't get vin's previous transaction")
+		}
+		for _, out := range prevQtumTx.Vouts {
+			for _, address := range out.Details.Addresses {
+				return utils.AddHexPrefix(address), nil
+			}
+		}
+	}
+	return "", errors.New("not found")
+}
+
+// NOTE:
+// 	- is not for reward transactions
+// 	- returning address already has 0x prefix
+//
+// 	TODO: researching
+// 	- Vout[0].Addresses[i] != "" - temporary solution
+func findNonContractTxReceiverAddress(vouts []*qtum.DecodedRawTransactionOutV) (string, error) {
+	for _, vout := range vouts {
+		for _, address := range vout.ScriptPubKey.Addresses {
+			if address != "" {
+				return utils.AddHexPrefix(address), nil
+			}
+		}
+	}
+	return "", errors.New("not found")
+}
+
+func getBlockNumberByHash(p *qtum.Qtum, hash string) (uint64, error) {
+	block, err := p.GetBlock(hash)
+	if err != nil {
+		return 0, errors.WithMessage(err, "couldn't get block")
+	}
+	return uint64(block.Height), nil
+}
+
+func getTransactionIndexInBlock(p *qtum.Qtum, txHash string, blockHash string) (int64, error) {
+	block, err := p.GetBlock(blockHash)
+	if err != nil {
+		return -1, errors.WithMessage(err, "couldn't get block")
+	}
+	for i, blockTx := range block.Txs {
+		if txHash == blockTx {
+			return int64(i), nil
+		}
+	}
+	return -1, errors.New("not found")
+}
+
+func formatQtumNonce(nonce int) string {
+	var (
+		hexedNonce     = strconv.FormatInt(int64(nonce), 16)
+		missedCharsNum = 16 - len(hexedNonce)
+	)
+	for i := 0; i < missedCharsNum; i++ {
+		hexedNonce = "0" + hexedNonce
+	}
+	return "0x" + hexedNonce
+}
+
+// Returns Qtum block number. Result depends on a passed raw param. Raw param's slice of bytes should
+// has one of the following values:
+// 	- hex string representation of a number of a specific block
+// 	- string "latest" - for the latest mined block
+// 	- string "earliest" for the genesis block
+// 	- string "pending" - for the pending state/transactions
+func getBlockNumberByParam(p *qtum.Qtum, rawParam json.RawMessage, defaultVal int64) (*big.Int, error) {
+	if len(rawParam) < 1 {
+		return nil, errors.Errorf("empty parameter value")
+	}
+	if !isBytesOfString(rawParam) {
+		return nil, errors.Errorf("invalid parameter format - string is expected")
+	}
+
+	param := string(rawParam[1 : len(rawParam)-1]) // trim \" runes
+	switch param {
+	case "latest":
+		res, err := p.GetBlockChainInfo()
+		if err != nil {
+			return nil, err
+		}
+		return big.NewInt(res.Blocks), nil
+
+	case "earliest":
+		// TODO: discuss
+		// ! Genesis block cannot be retreived
+		return big.NewInt(0), nil
+
+	case "pending":
```

Comment:
we don't really have the concept of pending when it comes to events.

Reply:
Are there any developments on the way in Qtum for this concept? We think that it could become an issue in the future, if it is not included.

Reply:
There are not. If you have concerns please detail them.